src/main/kotlin/io/zerobase/smarttracing/resources/PdfGenerator.kt

```diff
@@ -0,0 +1,131 @@
+        templateEngine.setTemplateResolver(templateResolver)
+        val data = organizationData(businessname, town, state)
+        val context = Context()
+        val dataMap = HashMap<String, Any>()
+        dataMap["data"] = data
+        dataMap["logoPath"] = QR_CODE_IMAGE_PATH
+        context.setVariables(dataMap)
+        // Flying Saucer needs XHTML - not just normal HTML. To make our life
+// easy, we use JTidy to convert the rendered Thymeleaf template to
+// XHTML. Note that this might not work for very complicated HTML. But
+// it's good enough for a simple letter.
+
+        val renderedHtmlContent = templateEngine.process("template", context)
+        val xHtml = convertToXhtml(renderedHtmlContent)
+        val renderer = ITextRenderer()
+        // FlyingSaucer has a working directory. If you run this test, the working directory
+// will be the root folder of your project. However, all files (HTML, CSS, etc.) are
+// located under "/src/test/resources". So we want to use this folder as the working
+// directory.
+        val baseUrl = FileSystems
+                .getDefault()
+                .getPath("src", "main", "resources", "pdfconfig")
+                .toUri()
+                .toURL()
+                .toString()
+        print(baseUrl);
+        renderer.setDocumentFromString(xHtml, baseUrl)
+        renderer.layout()
+        // And finally, we create the PDF:
+        val outputStream: OutputStream = FileOutputStream(OUTPUT_FILE)
+        renderer.createPDF(outputStream)
+        outputStream.close()
+    }
+
+    @Throws(UnsupportedEncodingException::class)
+    private fun convertToXhtml(html: String): String {
+        val tidy = Tidy()
+        tidy.inputEncoding = UTF_8
+        tidy.outputEncoding = UTF_8
+        tidy.xhtml = true
+        val inputStream = ByteArrayInputStream(html.toByteArray(charset(UTF_8)))
+        val outputStream = ByteArrayOutputStream()
+        tidy.parseDOM(inputStream, outputStream)
+        return outputStream.toString(UTF_8)
+    }
+
+    private fun organizationData(businessname: String, state: String, town: String): Data {
+        val data = Data()
+        data.businessname = businessname
+        data.state = state
+        data.town = town
+        return data
+    }
+
+    internal class Data {
+        var businessname: String? = null
+        var town: String? = null
+        var state: String? = null
+
+    }
+
+
+
+
+    companion object {
+        private const val OUTPUT_FILE = "src/main/resources/pdfconfig/pdfs/zerobase-qr.pdf"
+        private const val UTF_8 = "UTF-8"
+        private const val QR_CODE_IMAGE_PATH = "src/main/resources/pdfconfig/qr/processed/qr_logo_overlay.png"
+        private const val ZB_LOGO_IMAGE_PATH = "src/main/resources/pdfconfig/qr/rawfiles/zerobase_qr_logo.png"
+        private const val QR_VALUE = "https://zerobase.io/"
+        @Throws(Exception::class)
+        @JvmStatic
+        fun main(args: Array<String>) {
```

Comment:
this was ok for testing purposes, however for the real world situation we should not do this, i would like the pdf gen to output an input stream that we can use to attach to the email, and the email should call the generator function

Reply:
a) what does that have to do with a main function for testing?
b) it should return byte[] not a stream

Reply:
I haven't had a chance to take a look at email stuff. Referencing this right now: https://vangjee.wordpress.com/2010/11/02/how-to-create-an-in-memory-pdf-report-and-send-as-an-email-attachment-using-itext-and-java/
Reasoning to use byte array:
https://stackoverflow.com/questions/4817083/send-e-mail-in-pdf-attachment-as-stream

I've converted the function to return a byte array